src/functional/program_processors/codegens/gtfn/gtfn_ir_to_gtfn_im_ir.py

```diff
@@ -0,0 +1,274 @@
+from typing import Any, Dict, Iterable, List, Optional, Union
+
+import eve
+from eve import NodeTranslator, NodeVisitor
+from eve.utils import UIDGenerator
+from functional.program_processors.codegens.gtfn import gtfn_ir, gtfn_ir_common
+from functional.program_processors.codegens.gtfn.gtfn_im_ir import (
+    AssignStmt,
+    Conditional,
+    ImperativeFunctionDefinition,
+    InitStmt,
+    ReturnStmt,
+    Stmt,
+)
+
+
+def _find_connectivity(reduce_args: Iterable[gtfn_ir_common.Expr], offset_provider):
+    connectivities = []
+    for arg in reduce_args:
+        if isinstance(arg, gtfn_ir.FunCall) and arg.fun == gtfn_ir_common.SymRef(id="shift"):
+            assert isinstance(arg.args[-1], gtfn_ir.OffsetLiteral), f"{arg.args}"
+            connectivities.append(offset_provider[arg.args[-1].value])
+
+    if not connectivities:
+        raise RuntimeError("Couldn't detect partial shift in any arguments of reduce.")
+
+    if len({(c.max_neighbors, c.has_skip_values) for c in connectivities}) != 1:
+        # The condition for this check is required but not sufficient: the actual neighbor tables could still be incompatible.
+        raise RuntimeError("Arguments to reduce have incompatible partial shifts.")
+    return connectivities[0]
+
+
+def _is_reduce(node: gtfn_ir.FunCall):
+    return isinstance(node.fun, gtfn_ir.FunCall) and node.fun.fun == gtfn_ir_common.SymRef(
+        id="reduce"
+    )
+
+
+# the current algorithm to turn gtfn_ir into a imperative representation is rather naive,
+# it visits the gtfn_ir tree eagerly depth first, and it is only doing one pass. This entails
+# that some constructs can not be treated. This visitor checks for such constructs, and makes
+# the contract on the gtfn_ir for ToImpIR to suceed explicit
+@dataclasses.dataclass
+class IsImpCompatible(NodeVisitor):
+    incompatible_node: Optional[gtfn_ir_common.Expr] = None
+    compatible: bool = True
+
+    def visit_FunCall(self, node: gtfn_ir.FunCall, **kwargs):
+        if not self.compatible:
+            return
+
+        # reductions need at least one argument with a partial shift
+        if _is_reduce(node):
+            offset_provider = kwargs["offset_provider"]
+            assert offset_provider is not None
+            try:
+                _find_connectivity(node.args, offset_provider)
+            except RuntimeError:
+                self.compatible = False
+                self.incompatible_node = node
+
+        # function calls need to be lambdas, built ins or reduce
+        #   this essentailly avoids situations like
+        #   (λ(cs) → cs(w))(λ(x) → x)
+        #   where w is some field
+        if isinstance(node.fun, gtfn_ir_common.SymRef):
+            self.compatible = self.compatible and node.fun.id in gtfn_ir.BUILTINS
+            if not self.compatible:
+                self.incompatible_node = node
+
+        self.generic_visit(node, **kwargs)
+
+
+@dataclasses.dataclass(frozen=True)
+class ToImpIR(NodeVisitor):
```

Comment:
Good question. To me it wasn't a node translator because I'm not interested in translating the AST but I'm writing individual statements to the "tape". But yeah, will change that to a `NodeTranslator`